### DISCUSSION TITLE
M48T08

### DISCUSSION THREAD

**Discussion by I0KPT** [TomNisbet/TommyPROM#60] [General]
I would like to use TommyPROM to write a M48T08 BBRAM.  It's a 5 V 64 kbit SRAM incorporating a battery and a clock.  Has anyone already done it? Can I use an already defined PromDeviceXX class, or I need to create a new one?
Regards
Piero

**Comment by TomNisbet** [comment]
After a quick look at the data sheet, it looks like you could program the M48T08 using the default 28C driver.  You just need to set the block size to zero so the code writes a byte at a time instead of doing block writes.  In TommyPROM.ino, just comment out the first PromDevice28C declaration and uncomment the second one.  This is for the 28C64 chip, but it matches the need of your chips as well.

On the hardware side, you have 2 extra pins to deal with.  Connect the E2 pin to 5V so it is always enabled.  The Interrupt pin is an output and isn't needed, but the data sheet mentions it needing a pull-up resistor, so you may need to look into that.  Otherwise, just wire it up like a 28C chip, with the E1, G, and W pins as CE, OE, and WE, respectively.

**Reply by I0KPT** [reply]
Thank you, Tommy!   I am sure that your suggestion will help also other people interested in M48T08. It's a BBRAM used on some old laboratory instruments. In my case, it's on a Motorola MVME162 board, which is inside an Anritsu Microwave instrument.

**Reply by TomNisbet** [reply]
Please let me know how it goes and if you run into any problems.  If you are successful, I'll add it to the documentation.

**Reply by I0KPT** [reply]
Sure, I will inform you.   Probably I won't be able to test it before about a couple of weeks. Thank you again.

**Reply by I0KPT** [reply]
Tom, I just posted a comment on my first experiment with M48T08 and your programmer.

**Comment by I0KPT** [comment]
Hi Tom,

today I had some time to try to program M48T08 using your project.  I set-up everything on a breadboard and I am using an Arduino Nano.

I have some troubles.  Look at the two log file I am attaching at the end of this message. Which may be the problem of such a strange behavior?

I have compiled the TommyPROM.ino with the following lines:

    // Declare a global PROM device depending on the device type that is
    // defined in Configure.h
    #if defined(PROM_IS_28C)
    // Define a device for a 28C256 EEPROM with the following parameters:
    //   32K byte device capacity
    //   64 byte block writes
    //   10ms max write time
    //   Data polling supported
    //PromDevice28C  prom(32 * 1024L, 64, 10, true);
    PromDevice28C  prom(8 * 1024L, 0, 10, true);  // 28C64 with no page writes


Thank you a lot for your assistance!

best regards

Piero

>>>>> First Log

P00000 37 37 37

INFO: Poke successful

>P00000 37 37 37

INFO: Poke successful

>D00000
00000: 37 37 37 01  01 01 01 01  01 01 01 01  01 01 01 01  777..... ........
00010: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
00020: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
00030: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
00040: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
00050: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
00060: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
00070: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
00080: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
00090: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
000a0: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
000b0: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
000c0: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
000d0: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
000e0: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
000f0: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........

>P00000 38 38 38

INFO: Poke successful

>D00000
00000: 38 38 38 00  00 00 00 00  00 00 00 00  00 00 00 00  888..... ........
00010: 00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  ........ ........
00020: 00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  ........ ........
00030: 00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  ........ ........
00040: 00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  ........ ........
00050: 00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  ........ ........
00060: 00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  ........ ........
00070: 00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  ........ ........
00080: 00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  ........ ........
00090: 00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  ........ ........
000a0: 00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  ........ ........
000b0: 00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  ........ ........
000c0: 00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  ........ ........
000d0: 00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  ........ ........
000e0: 00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  ........ ........
000f0: 00 00 00 00  00 00 00 00  00 00 00 00  00 00 00 00  ........ ........

>>>>> Second log

TommyPROM 3.3 - 28C series EEPROM

Valid commands are:
  Bsssss eeeee    - Check to see if device range is Blank/erased (all FF)
  Csssss eeeee    - Compute checksum from device
  Dsssss eeeee    - Dump bytes from device to terminal
  Esssss eeeee    - Erase address range on device (needed for some Flash)
  Fsssss eeeee dd - Fill block on device with fixed value
  L               - Lock (enable) device Software Data Protection
  Psssss dd dd... - Poke (write) values to device (up to 32 values)
  Rsssss eeeee    - Read from device and save to XMODEM CRC file
  U               - Unlock (disable) device Software Data Protection
  Wsssss          - Write to device from XMODEM CRC file

  I               - Print debug Info
  Ssssss eeeee    - Scan addresses (read each 10x)
  Tsssss          - Test read address (read 100x)
  Zsssss          - Zap (burn) a 32 byte test pattern

>Z00000

ERROR: Verify failed  addr=0x1  read=0x43  expected=0x42

>
>Z00000

ERROR: Verify failed  addr=0x1  read=0x43  expected=0x42

>D00000
00000: 41 43 43 45  45 47 47 49  01 03 05 09  11 21 41 81  ACCEEGGI .....!A.
00010: 7f bf df ef  f7 fb fd ff  01 ff 55 ab  31 31 33 33  ........ ..U.1133
00020: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
00030: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
00040: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
00050: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
00060: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
00070: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
00080: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
00090: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
000a0: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
000b0: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
000c0: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
000d0: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
000e0: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........
000f0: 01 01 01 01  01 01 01 01  01 01 01 01  01 01 01 01  ........ ........

**Comment by I0KPT** [comment]
Hi Tom,

look also at the attached log file. Please note that the last bytes are updated automatically by the clock, which is inside the BBRAM.  I first filled all the BBRAM by "22". Then I poke P00000 01. The byte 0x0 was correctly updated, but all other bytes now are (or appear) "23".  

Best regards

Piero
[teraterm3.log](https://github.com/TomNisbet/TommyPROM/files/14226870/teraterm3.log)

**Reply by TomNisbet** [reply]
It looks like maybe your connection to data pin zero of the RAM is open or shorted.  The values read seem to follow the last value you've written to that bit.  Double check your wiring between the Arduino and the RAM for the least significant data bit.  I've noticed that the Arduino often reads back that same value you last wrote if nothing is connected to the data lines, like when there is no target EEPROM in the burner.

Try poking multiple different values in and I'll bet the LSB will follow the last value written.
For example, 
```
f 0 f 00
d 0 f
p 0 00 ff
d 0 f
p 0 ff 00
d 0 f
```

**Comment by I0KPT** [comment]
Pictures are here.

Il Dom 11 Feb 2024, 12:04 Piero Tognolatti ***@***.***> ha
scritto:
